test/unit/Server/Middleware/TransactionMarkerMiddlewareTest.php

```diff
@@ -0,0 +1,109 @@
+use Ingenerator\CloudTasksWrapper\TestHelpers\Server\TestTaskHandler;
+use Ingenerator\PHPUtils\DateTime\DateString;
+use PHPUnit\Framework\TestCase;
+use test\mock\Ingenerator\CloudTasksWrapper\Repository\ArrayTransactionMarkerRepository;
+
+class TransactionMarkerMiddlewareTest extends TestCase
+{
+    private ArrayTransactionMarkerRepository $transaction_marker_repo;
+
+    public function test_it_is_initialisable()
+    {
+        $this->assertInstanceOf(TransactionMarkerMiddleware::class, $this->newSubject());
+        $this->assertInstanceOf(TaskHandlerMiddleware::class, $this->newSubject());
+    }
+
+    public function test_it_returns_response_of_next_in_chain_if_no_x_transaction_header()
+    {
+        $res = new ArbitraryTaskResult('no-transaction-header');
+        $req = TaskRequestStub::any();
+        $this->assertSame(
+            $res,
+            $this->newSubject()->process(
+                $req,
+                TestTaskChain::withHandler(TestTaskHandler::expectsReqAndReturns($req, $res))
+            )
+        );
+    }
+
+    public function test_it_returns_response_of_next_in_chain_if_transaction_marker_exists()
+    {
+        $this->transaction_marker_repo = ArrayTransactionMarkerRepository::withUuids(['abcd-efgh']);
+
+        $req = TaskRequestStub::with(['headers' => ['X-Transaction' => 'abcd-efgh']]);
+        $res = new ArbitraryTaskResult('transaction-marker-exists');
+        $this->assertSame(
+            $res,
+            $this->newSubject()->process(
+                $req,
+                TestTaskChain::withHandler(TestTaskHandler::expectsReqAndReturns($req, $res))
+            )
+        );
+    }
+
+    public function test_it_does_not_run_if_request_is_before_expiry()
+    {
+        $future = DateString::ymdhis(new \DateTimeImmutable('tomorrow'));
+
+        $request = TaskRequestStub::with(
+            ['headers' => ['X-Transaction' => 'abcd-1234', 'X-Transaction-Expire' => $future]]
+        );
+
+        $this->assertSame(
+            [
+                'code'        => TransactionMarkerResult::NOT_YET_VISIBLE,
+                'msg'         => 'transaction not ready',
+                'log_context' => [
+                    'uuid'   => 'abcd-1234',
+                    'expiry' => $future,
+                ],
+            ],
+            $this->newSubject()->process($request, TestTaskChain::nextNeverCalled())->toArray()
+        );
+    }
+
+    public function test_it_does_not_run_if_after_expiry()
+    {
+        $request = TaskRequestStub::with(
+            ['headers' => ['X-Transaction' => 'abcd-1234', 'X-Transaction-Expire' => '2021-02-04 01:02:03']]
+        );
+
+        $this->assertSame(
+            [
+                'code'        => TransactionMarkerResult::TRANSACTION_EXPIRED,
+                'msg'         => 'transaction rolled back',
+                'log_context' => [
+                    'uuid'   => 'abcd-1234',
+                    'expiry' => '2021-02-04 01:02:03',
```

Comment:
Minor stylistic thing, I prefer to consistently use either stubbed time or real time. E.g. if the class is using RealtimeClock then use fixed time strings for "now" and the test value. 

But in this case as the class under test is using `new DatetimeImmutable` then the test value should be eg `Datestring::ymdhis(new DatetimeImmutable('-1s'))`.

Not really any way it can make a difference with something this simple but it helps make sure we're testing boundary conditions and avoids risk of oddities.